libnmstate/appliers/linux_bridge.py

```diff
@@ -29,3 +38,51 @@ def set_bridge_ports_metadata(master_state, slave_state):
     ports = master_state.get("bridge", {}).get("port", [])
     port = next(filter(lambda n: n["name"] == slave_state["name"], ports), {})
     slave_state["_brport_options"] = port
+
+
+def generate_port_vlan_metadata(desired_state, current_state):
```

Comment:
For this PR, I don't think we need to use the metadata.

The `bridge.create_setting()` could determine the set `bridge_setting.props.vlan_filtering` or not
based on `iface_state` passing to it.

This could save code editing on `metadata.py` and `appliers/linux_bridge.py` but limit it inside of `nm/bridge.py`.

Reply:
AFAIK, that wouldn't be enough, since we need information from the current_state. It would true *if* we had exposed the boolean that activates vlan filtering.

That leaves metadata or state_merge as the only options (since that's where we have access to current_state & desired_state.

I chose metadata, since it feels more self-contained.

Reply:
At time of `nm.bridge.create_setting()` been called. The state has been merged, hence you don't need to worry about current state.

Reply:
I am expecting something like:

```

def create_setting(iface_state, base_con_profile):
    bridge_setting = _get_current_bridge_setting(base_con_profile)
    if not bridge_setting:
        bridge_setting = nmclient.NM.SettingBridge.new()

    bridge_setting.props.vlan_filtering = _have_port_vlan(iface_state)


def _has_port_vlan(iface_state):
    """
    Return True if iface_state contains:
       [LB.CONFIG_SUBTREE][LB.PORT_SUBTREE][LB.Port.VLAN_SUBTREE]
    """
    return any(
        LB.Port.VLAN_SUBTREE in port
        for port in iface_state.get(LB.CONFIG_SUBTREE, {}).get(
            LB.PORT_SUBTREE, []
        )
    )
```

The metadata is for information referring to other interface, not for this case.
This is the comment of `generate_ifaces_metadata()` which explained:
```
    The described desired state for each interface may include references to
    other interfaces. As the provider handles the interface setting in an
    isolated manner, it is sometime necessary to specify a property on an
    interface based on a property from a different interface.
    An example of this, is the bond slaves property, which should mark each
    slave with the setting of it being a slave.
```

Reply:
Have you considered @EdDev 's comment on https://github.com/nmstate/nmstate/pull/576#discussion_r355195141 ? 

I quote: "I am also unsure if this check is even correct. Bridge options may be changed without specifying any vlans, but that is not to say that there are no current such ports defined. (unless you handle this as part of the "merge" step)."

From your excerpt, if you update the bridge options of - let's say - a bridge that has vlan filtering active, but don't update the port vlan configuration of any of it's ports, it'll disable the bridge's vlan filtering, wouldn't it ?

What you're suggesting is literally what I had in the first place (well, my code was located on the applier, not the bridge module, but anyway... ).

Reply:
@EdDev Got it wrong. The vlan filter status can be solely determined by `iface_state`, hence it should not use metadata.

The reason you noticed the `nm.bridge.create_settting()` only get partial data on port section is because the
`state.dict_update()` only do full copy on list.

So the the correct approach should be:
 * Amend the `state.merge_interfaces()` to merge the port settings from current to desire.
 * Passing the merged interface state to `bridge.create_setting()` where it could determine to enable vlan filtering or not.

This is the patch based on your latest commit b71e471:

https://gist.github.com/cathay4t/1dabb3074e1dc003afbd0619254ad12e

I tested it with the yaml files https://paste.centos.org/view/8dad1320

This PR has 100+ comments and out of control. I would suggest you to split this PR into:

 * PR for getting run time status. Reading on-disk config is not runtime status.
   As NM does not provides so, please try to parse `bridge -j vlan show` outputs.

 * PR for create bridge setting and bridge port setting in `libnmstate.nm` folder only.
   I am OK if you don't want to waste time on creating integration test case for `libnmstate.nm` files.
 * PR for amend the `state.py` to merge bridge port settings.
 * PR for full support of linux bridge vlan filtering including integration test cases.
 * PR for adding example of linux bridge vlan filtering.

Reply:
I've set up a meeting w/ you, where I hope we can discuss most of this.

Please accept or propose another time.

On another note, I have one question about your 'merge_interfaces' proposal;  how can port vlan filtering be disabled on a port that currently has it configured ?

I've always assumed that under this model (where vlan filtering being enabled is not exposed)  , disabling it would be done if the user referred in the desired state a port that currently has vlan filtering, without indicating it's vlan config. 

I'm wondering how can that be done using the merge interfaces, since, afaiu, it's purpose is to (and I quote the comment from the code): "Given the self and other states, complete the self state by merging the missing parts from the current state" .

Reply:
> PR for getting run time status. Reading on-disk config is not runtime status. As NM does not provides so, please try to parse bridge -j vlan show outputs.

We have a bad experience with `bridge -j vlan` from kubernetes-nmstate. It is not always a valid JSON and the output format varies on the OS you run on. If possible, I would advice against using it, at least in production code.

Reply:
> > PR for getting run time status. Reading on-disk config is not runtime status. As NM does not provides so, please try to parse bridge -j vlan show outputs.
> 
> We have a bad experience with `bridge -j vlan` from kubernetes-nmstate. It is not always a valid JSON and the output format varies on the OS you run on. If possible, I would advice against using it, at least in production code.

Thanks for the info.

Reply:
> how can port vlan filtering be disabled on a port that currently has it configured ?

Need to set clear rule on how things been merged from current. My propose would be:
 * When desire state does not contain port section, it will copy from current state.
 * When desire state has port section defined as empty list. New configure will remove all  the ports from bridge.
 * When desire state has port section with name only. All information for that port will copy fro current state.
 * When desire state has port section with properties more than name, __no__ information will be copy from current state and let nmstate determine the default value for undefined properties. This means, when user want to change the setting of certain port, it have to define all setting of that port in desire state and expect no merge.

To answer your question, to support that use case, we just ask user to provide full configure
of that port without vlan filter settings. When all the port does not have vlan filter, the bridge will disable vlan filter.

Reply:
interface merging PR: https://github.com/nmstate/nmstate/pull/888